### PR TITLE
[SYCL] Enable nonsemantic.shader.debuginfo.200 by default

### PR DESCRIPTION
**Pull request by MrSidims**
It's left disabled only for FPGA target.

### CODE REVIEW

File: clang/test/Driver/sycl-spirv-default-options.c [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Generate .o file as SYCL device library file.
+//
+// RUN: touch %t.devicelib.cpp
+// RUN: %clang %t.devicelib.cpp -fsycl -fsycl-targets=spir64-unknown-unknown -c --offload-new-driver -o %t_1.devicelib.o
+// RUN: %clang %t.devicelib.cpp -fsycl -fsycl-targets=spir64_gen-unknown-unknown -c --offload-new-driver -o %t_2.devicelib.o
+// RUN: %clang %t.devicelib.cpp -fsycl -fsycl-targets=spir64_x86_64-unknown-unknown -c --offload-new-driver -o %t_3.devicelib.o
+
+// Test for default llvm-spirv options
```
Comment (MrSidims) [inline clang/test/Driver/sycl-spirv-default-options.c:8]: I haven't found the proper tests for this, so created a new one. The idea of it that when we decide to support new SPIR-V versions etc - we could modify this tests. Similar test is sycl-spirv-ext.c - but it was created for a bit different purpose and I'd like not to modify it with this patch.

Comment (mdtoguchi) [reply clang/test/Driver/sycl-spirv-default-options.c:8]: This test looks to cover the new offloading model with the changes for `clang-linker-wrapper`.  Should an equivalent test be created for the old offloading model behaviors to check the driver defaults?

Comment (MrSidims) [reply clang/test/Driver/sycl-spirv-default-options.c:8]: added a new test, thanks!

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp [hunk 2]
```diff
@@ -830,19 +830,7 @@ getTripleBasedSPIRVTransOpts(const ArgList &Args,
...
-  if (EnableNonSemanticDebug) {
-    TranslatorArgs.push_back(
-        "-spirv-debug-info-version=nonsemantic-shader-200");
-  } else {
-    TranslatorArgs.push_back("-spirv-debug-info-version=ocl-100");
-    // Prevent crash in the translator if input IR contains DIExpression
-    // operations which don't have mapping to OpenCL.DebugInfo.100 spec.
-    TranslatorArgs.push_back("-spirv-allow-extra-diexpressions");
-  }
+  TranslatorArgs.push_back("-spirv-debug-info-version=nonsemantic-shader-200");
```
Comment (sarnex) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:833]: It looks like in the driver we don't want to do this for FPGA mode, and I don't see that implemented here, and I see at least one check for FPGA in this file:

```
Triple.getSubArch() != llvm::Triple::SPIRSubArch_fpga
```
Should we add that here?

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -10674,12 +10674,8 @@ static void getTripleBasedSPIRVTransOpts(Compilation &C,
                                          ArgStringList &TranslatorArgs) {
   bool IsCPU = Triple.isSPIR() &&
                Triple.getSubArch() == llvm::Triple::SPIRSubArch_x86_64;
-  // Enable NonSemanticShaderDebugInfo.200 for CPU AOT and for non-Windows
-  const bool IsWindowsMSVC =
-      Triple.isWindowsMSVCEnvironment() ||
-      C.getDefaultToolChain().getTriple().isWindowsMSVCEnvironment();
-  const bool EnableNonSemanticDebug =
-      IsCPU || (!IsWindowsMSVC && !C.getDriver().IsFPGAHWMode());
+  // Enable NonSemanticShaderDebugInfo.200 for non-FPGA targets.
```
Comment (sarnex) [inline clang/lib/Driver/ToolChains/Clang.cpp:10677]: This shouldn't be ABI breaking right? As in an will an object file compiled before this change still work?

Comment (MrSidims) [reply clang/lib/Driver/ToolChains/Clang.cpp:10677]: > This shouldn't be ABI breaking right? As in an will an object file compiled before this change still work?

Yes, we don't deprecate OpenCL.DebugInfo.100 , our drivers still support it. Please also note, that New debug info format was already enabled for Linux ~8 months ago. Right now I'm enabling it for Windows as well.

Comment (sarnex) [reply clang/lib/Driver/ToolChains/Clang.cpp:10677]: cool, thx

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp (hunk 2 above)
Comment (MrSidims) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:833]: I can add it. Also the current version of `getTripleBasedSPIRVTransOpts` in ClangLinkerWrapper.cpp doesn't consider FPGA, while the logic must be exactly like in https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/ToolChains/Clang.cpp#L10728 . This should go into a separate patch.

Comment (sarnex) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:833]: yeah thats fine for now, i guess lets just make an internal tracker to decide what we're doing with fpga

**Review by sarnex — APPROVED** [review]

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp (hunk 2 above)
Comment (asudarsa) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:833]: Thanks @sarnex. There are some unnecessary FPGA related logic which has snuck into the wrapper. That needs to be cleaned up. In my todo list. :-)

Comment (sarnex) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:833]: thx!

**Review by mdtoguchi — APPROVED** [review]
LGTM

**Pull request merged**